native/thrift_struct_read_module.c

```diff
@@ -0,0 +1,1851 @@
+        PyDict_SetItemString(globalVars,"contextType",Py_BuildValue("i",contextType));
+        PyDict_SetItemString(globalVars,"first",Py_BuildValue("i",first));
+    }
+    else {
+        PyDict_SetItemString(globalVars,"contextType",Py_BuildValue("i",BASECONTEXT));
+        PyDict_SetItemString(globalVars,"first",Py_BuildValue("i",1));
+    }
+}
+
+int
+JTypesToInt(char *CharTypes)
+{
+    if (!strcmp(CharTypes, "tf")) {
+        return BOOL;
+    }
+    else if (!strcmp(CharTypes, "i8")) {
+        return BYTE;
+    }
+    else if (!strcmp(CharTypes, "dbl")) {
+        return DOUBLE;
+    }
+    else if (!strcmp(CharTypes, "i16")) {
+        return I16;
+    }
+    else if (!strcmp(CharTypes, "i32")) {
+        return I32;
+    }
+    else if (!strcmp(CharTypes, "i64")) {
+        return I64;
+    }
+    else if (!strcmp(CharTypes, "str")) {
+        return STRING;
+    }
+    else if (!strcmp(CharTypes, "rec")) {
+        return STRUCT;
+    }
+    else if (!strcmp(CharTypes, "map")) {
+        return MAP;
+    }
+    else if (!strcmp(CharTypes, "set")) {
+        return SET;
+    }
+    else if (!strcmp(CharTypes, "lst")) {
+        return LIST;
+    }
+    else {
+        return -1;
+    }
+}
+
+/*TCompact C functions*/
+int
+readSize(char *data)
+{
+    int result;
+
+    result = readVarint(data);
+    if (result < 0) {
+        PyErr_SetString(PyExc_ValueError, "Length <0");
+        return -1;
+    }
+    return result;
+}
+
+int
+readVarint(char *data)
+{
+    int result;
+    int shift;
+    int x;
+
+    result = 0;
+    shift = 0;
+    while (1) {
+        x = readByteC(data);
+        result = result | ((x & 127)<<shift);
+        if (x>>7 == 0) {
+            return result;
+        }
+        shift = shift+7;
+    }
+}
+
+int
+fromZigZag(int n)
+{
+    return (n>>1)^-(n&1);
+}
+
+int
+readerZigZag(char *data)
+{
+    int state;
+    PyObject *globalVars;
+    PyObject *py_state;
+
+    globalVars = PyThreadState_GetDict();
+    py_state = PyDict_GetItemString(globalVars,"state");
+    if (!PyArg_Parse(py_state, "i", &state)) {
+        return ;
+    }
+
+    if (state != VALUE_READ && state != CONTAINER_READ) {
+        PyErr_SetString(PyExc_ValueError, "wrong estate");
+        return -1;
+    }
+    return fromZigZag(readVarint(data));
+}
+
+int
+getTType(int byte)
+{
+    int type;
+
+    type = byte & 15;
+    if (type == 0) {
+        return STOP;
+    }
+    if (type == 1) {
+        return BOOL;
+    }
+    if (type == 2) {
+        return BOOL;
+    }
+    if (type == 4) {
+        return I16;
+    }
+    if (type == 5) {
+        return I32;
+    }
+    if (type == 6) {
+        return I64;
+    }
+    if (type == 7) {
+        return DOUBLE;
+    }
+    if (type == 8) {
+        return STRING;
+    }
+    if (type == 9) {
+        return LIST;
+    }
+    if (type == 10) {
+        return SET;
+    }
+    if (type==11) {
+        return MAP;
+    }
+    if (type == 12) {
+        return STRUCT; 
+    }
```

Comment:
nit: it'll be more readable to use a switch/case statement here instead of all the ifs